source/manual/test-and-build-a-project-with-github-actions.html.md

```diff
@@ -0,0 +1,266 @@
+
+Notes:
+
+- [ruby/setup-ruby](https://github.com/ruby/setup-ruby) will automatically
+  select the Ruby version based on a `.ruby-version` file and will install
+  bundler if a necessary.
+- This use [actions/cache](https://github.com/actions/cache) to cache gems
+  between runs.
+- This uses the [deployment][bundler-deployment] flag for bundler to be
+  configured for CI environments, the jobs flag allows parallel installs and
+  retries allows re-attempts on failures.
+
+### A Ruby Gem
+
+This differs to a simple Ruby application as Ruby gems are not tied to a
+particular Ruby version, therefore for those we should test a variety of
+Ruby versions. Do bear in mind that due to the restriction on secrets
+usage we cannot currently publish gems through GitHub Actions, this can
+still be achieved through Jenkins.
+
+```yml
+# .github/workflows/ci.yml
+name: Continuous Integration
+on: [push, pull_request]
+jobs:
+  test:
+    strategy:
+      matrix:
+        ruby: [2.5, 2.6, 2.7]
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: ruby/setup-ruby@v1
+        with:
+          ruby-version: ${{ matrix.ruby }}
+      - uses: actions/cache@v1
+        with:
+          path: vendor/bundle
+          key: bundle-${{ matrix.ruby }}-${{ hashFiles('**/*.gemspec') }}
+          restore-keys: bundle-${{ matrix.ruby }}
+      - run: bundle install --jobs 4 --retry 3 --path vendor/bundle
+      - run: bundle exec rake
+```
+
+Notes:
+
+- The [deployment][bundler-deployment] flag for bundler cannot be used for gems
+  as they should not have a Gemfile.lock committed.
+- Our preference is to test the gems we publish against the latest version of
+  [all currently supported minor versions of Ruby MRI][ruby-branches].
+
+## GOV.UK Rails application with Postgres, Redis, Yarn and GOV.UK Content Schemas dependencies
+
+This is an example that suits a relatively complex GOV.UK Rails application
+(this was written for [Content Publisher][]) that has software dependencies
+(provided by [Docker containers][actions-docker-services]), cloning of a
+supplementary repository, and the use of Node.js for JavaScript dependency
+management.
+
+```yml
+name: Continuous Integration
+on: [push, pull_request]
+jobs:
+  test:
+    runs-on: ubuntu-latest
+    services:
+      postgres:
+        image: postgres:9.4
+        ports:
+          - 5432:5432
+        env:
+          POSTGRES_HOST_AUTH_METHOD: trust
+        options: --health-cmd pg_isready --health-interval 10s --health-timeout 5s --health-retries 5
+      redis:
+        image: redis
+        ports:
+          - 6379:6379
+        options: --health-cmd "redis-cli ping" --health-interval 10s --health-timeout 5s --health-retries 5
+    env:
+      RAILS_ENV: test
+      REDIS_URL: redis://localhost:6379/0
+      TEST_DATABASE_URL: postgresql://postgres@localhost/content-publisher
+      GOVUK_CONTENT_SCHEMAS_PATH: vendor/govuk-content-schemas
+    steps:
+      - name: Clone project
```

Comment:
```suggestion

```

This would be more consistent with the other examples. Perhaps we could just have names for the 'odd' action in a pair?

Reply:
I felt that things got a bit more complex if you only name some actions with the same name. In this example we've got the schemas as the odd example but then you could have a second or third. I don't mind too much but just would want to go with the route that involves the least decisions for an implementor.

Reply:
I suppose you end up making a decision either way, as you need to be aware that you're writing a flow with repeated actions. Since you can always expand each action to see what it's doing (e.g. which repo it's fetching), do the names really add much? I appreciate it's surprising to see something twice; my hunch is that in practice, people will just forget to add the names, since it requires more effort than not.

Reply:
Sure, I guess we'll see what people actually do in practice. I think it's stuff worth suggesting through conventions that identical steps should be named.

One thing to ponder is whether we should actually highlight a convention for naming steps. Thinking about it we might want to specify an overall convention for names since we have them at workflow level (we have a preference for no-name), job level (preference for no name as job key should be sufficient) and step (preference for no-name unless necessary to resolve ambiguity), which I think wrapped together sounds like everything has a preference for no-name unless resolving some ambiguity.

Reply:
An overall note on convention sounds good to me 👍. Still not convinced resolving ambiguities meets a need: if a step fails, then the name is pretty irrelevant, since the person debugging it needs to look into the detailed log anyway; if a step passes, then I wouldn't expect anyone to pay much attention to its name.

However, I'm not too fussed one way or the other.